node/core/dispute-coordinator/src/real/mod.rs

```diff
@@ -499,10 +502,212 @@ async fn handle_new_activations(
+}
+
+/// Scrapes on-chain votes (backing votes and concluded disputes) for a active leaf of the relay chain.
+async fn scrape_on_chain_votes(
+	ctx: &mut (impl SubsystemContext<Message = DisputeCoordinatorMessage>
+	          + overseer::SubsystemContext<Message = DisputeCoordinatorMessage>),
+	overlay_db: &mut OverlayedBackend<'_, impl Backend>,
+	state: &mut State,
+	new_leaf: Hash,
+	now: u64,
+	metrics: &Metrics,
+) -> Result<(), Error> {
+	// obtain the concluded disputes as well as the candidate backing votes
+	// from the new leaf
+	let ScrapedOnChainVotes { session, backing_validators, disputes } = {
+		let (tx, rx) = oneshot::channel();
+		ctx.send_message(RuntimeApiMessage::Request(
+			new_leaf,
+			RuntimeApiRequest::FetchOnChainVotes(tx),
+		))
+		.await;
+		match rx.await {
+			Ok(Ok(Some(val))) => val,
+			Ok(Ok(None)) => {
+				tracing::trace!(
+					target: LOG_TARGET,
+					relay_parent = ?new_leaf,
+					"No on chain votes stored for relay chain leaf");
+				return Ok(())
+			},
+			Ok(Err(e)) => {
+				tracing::debug!(
+					target: LOG_TARGET,
+					relay_parent = ?new_leaf,
+					error = ?e,
+					"Could not retrieve on chain votes due to an API error");
+				return Ok(())
+			},
+			Err(e) => {
+				tracing::debug!(
+					target: LOG_TARGET,
+					relay_parent = ?new_leaf,
+					error = ?e,
+					"Could not retrieve onchain votes due to oneshot cancellation");
+				return Ok(())
+			},
+		}
+	};
+
+	if backing_validators.is_empty() && disputes.is_empty() {
+		return Ok(())
+	}
+
+	// Obtain the session info, for sake of `ValidatorId`s
+	// either from the rolling session window.
+	// Must be called _after_ `fn cache_session_info_for_head`
+	// which guarantees that the session info is available
+	// for the current session.
+	let session_info: SessionInfo =
+		if let Some(session_info) = state.rolling_session_window.session_info(session) {
+			session_info.clone()
+		} else {
+			tracing::warn!(
+				target: LOG_TARGET,
+				relay_parent = ?new_leaf,
+				"Could not retrieve session info from rolling session window");
+			return Ok(())
+		};
+
+	// Scraped on-chain backing votes for the candidates with
+	// the new active leaf as if we received them via gossip.
+	for (candidate_receipt, backers) in backing_validators {
+		let candidate_hash = candidate_receipt.hash();
+		let statements =
+			backers.into_iter().filter_map(|(validator_index, validator_signature)| {
+				let validator_public: ValidatorId = session_info
+					.validators
+					.get(validator_index.0 as usize)
+					.or_else(|| {
+						tracing::error!(
+						target: LOG_TARGET,
+						relay_parent = ?new_leaf,
+						"Missing public key for validator {:?}",
+						&validator_index);
+						None
+					})
+					.cloned()?;
+				let signed_dispute_statement =
+					SignedDisputeStatement::new_unchecked_from_trusted_source(
+						// TODO verify this is the correct statement
+						DisputeStatement::Valid(ValidDisputeStatementKind::BackingValid(new_leaf)),
```

Comment:
Likely not correct. Both `BackingValid` and `BackingSeconded` statements can appear on-chain.

Reply:
We might piggyback on the disputes storage items instead, which contain the `DisputeStatement`.

Reply:
No, that doesn't help because that'd only store things once they're already disputed and this PR is about making sure that the validators all have the necessary 'valid' votes on bad blocks to begin a dispute during approval-checking.

Reply:
i.e. the attack scraping backing votes solves is about:
1. Malicious validator group submits backing votes to the chain without gossiping them
2. Approval checkers find that the block is bad, but don't have any 'valid' votes in their dispute DB to take the other side of the dispute
3. Approval checkers fail to initiate a dispute as they see it as one-sided and the bad block survives.

This PR inserts a step 1a: "all honest validators (including approval checkers) scrape the backing votes from the chain to ensure they have the opposing votes", and this makes step 2 of the attack impossible.

Reply:
I see, thanks for clarifying.